### PR TITLE
WIP: implement convolution

### PR DESCRIPTION
**Pull request by kohr-h**
Going a different route this time. First discrete convolution (`R^n -> R^n`) using direct or FFT methods, then discretized continuous convolution using the discrete one internally.

I chose ``issue-342__tensor`` as base rather than ``master``, since implementation was a bit easier. So this branch goes in after the tensor one.

**TODO:**
- [x] Implement  `DiscreteConvolution`
- [x] ~~Add option to provide kernel in Fourier space~~ This is too much of a hassle, makes the initializer way too complicated. It's better to expose helpers to do this manually.
- [x] Turn assertions into proper exceptions with error messages
- [x] Do all input checks according to the documentation
- [x] Write unit tests for `DiscreteConvolution`
- [x] Profile the code to see whether it has avoidable bottlenecks (no obvious ones)
- [x] Implement `DiscreteConvolution.adjoint`
- [ ] Optimize case when padding is 0, since `out` can be used then
- [ ] Write example for `DiscreteConvolution`
- [ ] Implement `Convolution`
- [ ] Write unit tests for `Convolution`
- [ ] Write example for `Convolution`

### CODE REVIEW

**Comment by kohr-h** [comment]
Note: I made `oplib` a place for "advanced" operators that cannot be part of any other subpackage due to circular dependencies. Currently in there are `ufunc_ops.py` and `convolution.py`. As discussed in another thread (#944), we can think about moving functionals to the same place.

File: odl/oplib/convolution.py [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+            Space of output elements of the convolution. Must be of the
+            same shape as ``domain``. If not given, the range is equal to
+            ``domain.astype(result_dtype)``, where ``result_dtype`` is
+            the data type of the convolution result. If ``impl='real'``,
+            integer dtypes are preserved, while for ``impl='fft'``,
+            the smallest possible floating-point type is chosen.
+        axis : int or sequence of ints, optional
+            Coordinate axis or axes in which to take the convolution.
+            ``None`` means all input axes.
+        impl : {'fft', 'real'}
```
Comment (kohr-h) [inline odl/oplib/convolution.py]: Not sure about naming, maybe `'direct'` would be better.

File: odl/oplib/convolution.py [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+            axis, with padding values corresponding to ``axis`` entries
+            as provided.
+            Default: ``min(kernel.shape - 1, 64)``
+        padded_shape : sequence of ints, optional
+            Apply zero-padding with this target shape. Cannot be used
+            together with ``padding``.
+        cache_kernel_ft : bool, optional
+            If ``True``, store the Fourier transform of the kernel for
+            later reuse.
+            Default: ``False``
```
Comment (kohr-h) [inline odl/oplib/convolution.py]: There should also be an option to provide the kernel directly in Fourier space.

File: odl/oplib/convolution.py [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+        ...                    [-3, -6, -9]])
+        >>> conv(x)
+        rn((3, 3)).element(
+            [[  1.,   3.,   5.],
+             [  3.,   9.,  15.],
+             [ -1.,  -3.,  -5.]]
+        )
+
+        Convolution in selected axes can be done either with broadcasting
+        or with "stacked kernels":
```
Comment (kohr-h) [inline odl/oplib/convolution.py]: Mention that this is not possible with `impl='real'`

**Comment by adler-j** [comment]
Why close this?

**Comment by kohr-h** [comment]
Got auto-closed when the base branch disappeared, will make a new PR into master.

**Pull request closed without merging**